### PR TITLE
implement density functions for terrain generation

### PR DESCRIPTION
**Pull request by kralverde**
WIP

This PR has the end goal of generating plains biomes

### CODE REVIEW

**Comment by kralverde** [comment]
I bit off more than I can chew. Here is another stepping stone towards chunk generation: the density functions. I will keep on working towards terrain generation.

File: pumpkin-core/src/math/mod.rs [hunk 1]
```diff
@@ -20,8 +21,20 @@ pub fn squared_magnitude(a: f64, b: f64, c: f64) -> f64 {
...
     coord >> 4
 }
+
+pub fn squared_hypot(a: f64, b: f64) -> f64 {
+    a * a + b * b
+}
+
+pub fn hypot(a: f64, b: f64) -> f64 {
+    squared_hypot(a, b).sqrt()
+}
```
Comment (Snowiiii) [inline pumpkin-core/src/math/mod.rs:40]: Rust already provides hypot functions, Why use own instead ?

File: pumpkin-world/src/world_gen/noise/density/end.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                    let r = 100f32 - (h * h + q * q).sqrt() * g;
+                    let s = r.clamp(-100f32, 80f32);
+
+                    f = f.max(s);
+                }
+            }
+        }
+
+        f
+    }
```
Comment (Snowiiii) [inline pumpkin-world/src/world_gen/noise/density/end.rs:55]: Is is vanilla ?

**Comment by Snowiiii** [comment]
I see you almost have `function1`, `function2` `function3`... everywhere. Did you got this code from somewhere or why var names are so weird?

**Comment by kralverde** [comment]
Straight from Java lol

File: pumpkin-world/src/world_gen/noise/density/end.rs (hunk 2 above)
Comment (kralverde) [reply pumpkin-world/src/world_gen/noise/density/end.rs:55]: Yes

File: pumpkin-core/src/math/mod.rs (hunk 1 above)
Comment (kralverde) [reply pumpkin-core/src/math/mod.rs:40]: I'll use rust's

**Comment by Snowiiii** [comment]
I don't like if someone just copies Java code and then even say that he don't know what the code does. So you said your goal is to have the Plains biome generation working. Do you need everything here for this, Or is something missing ?

**Comment by Snowiiii** [comment]
I would like to have good function names at least

**Comment by kralverde** [comment]
Functions have been renamed

**Comment by Snowiiii** [comment]
noice, can you may fix the conflicts ? 
then it should be ready to merge

**Comment by kralverde** [comment]
rebased, should be good

**Pull request merged**